be/src/storage/binlog_file_reader.cpp

```diff
@@ -0,0 +1,331 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "storage/binlog_file_reader.h"
+
+#include "fs/fs.h"
+#include "storage/binlog_file_writer.h"
+#include "storage/rowset/page_io.h"
+#include "util/coding.h"
+#include "util/crc32c.h"
+#include "util/raw_container.h"
+
+namespace starrocks {
+
+BinlogFileReader::BinlogFileReader(std::string file_name, std::shared_ptr<BinlogFileMetaPB> file_meta)
+        : _file_path(std::move(file_name)), _file_meta(file_meta) {}
+
+Status BinlogFileReader::seek(int64_t version, int64_t seq_id) {
+    std::shared_ptr<FileSystem> fs;
+    ASSIGN_OR_RETURN(fs, FileSystem::CreateSharedFromString(_file_path))
+    ASSIGN_OR_RETURN(_file, fs->new_random_access_file(_file_path))
+    ASSIGN_OR_RETURN(_file_size, _file->get_size())
+    Status status = _read_file_header();
+    if (!status.ok()) {
+        LOG(WARNING) << "Fail to parse file header " << _file_path << ", " << status;
+        return Status::InternalError("Fail to parse file header " + _file_path);
+    }
+    _next_page_index = 0;
+    _current_page_context = std::make_unique<PageContext>();
+    _current_log_entry = std::make_unique<LogEntryInfo>();
+
+    return _seek(version, seq_id);
+}
+
+Status BinlogFileReader::next() {
+    if (_is_end_of_file()) {
+        return Status::EndOfFile("There is no left entries");
+    }
+
+    PageContext* page_context = _current_page_context.get();
+    if (page_context->next_log_entry_index >= page_context->page_header.num_log_entries()) {
+        // end of current page and read the next
+        RETURN_IF_ERROR(_read_next_page());
+    }
+    _advance_log_entry();
+    return Status::OK();
+}
+
+LogEntryInfo* BinlogFileReader::log_entry() {
+   return _current_log_entry.get();
+}
+
+Status BinlogFileReader::_seek(int64_t version, int64_t seq_id) {
+    RETURN_IF_ERROR(_seek_to_page(version, seq_id));
+    RETURN_IF_ERROR(_seek_to_log_entry(seq_id));
+    return Status::OK();
+}
+
+Status BinlogFileReader::_seek_to_page(int64_t version, int64_t seq_id) {
+    _reset_current_page_context();
+    do {
+        _reset_current_page_context();
+        PageHeaderPB& page_header_pb = _current_page_context->page_header;
+        RETURN_IF_ERROR(_read_page_header(_next_page_index, &page_header_pb));
+        if (page_header_pb.version() > version) {
+            return Status::NotFound(strings::Substitute("Can't find version $0", version));
```

Comment:
`fmt::format` is better? Add `page_header_pb.version()` in the error msg to debug?